bin/filter_gtf.py

```diff
@@ -0,0 +1,119 @@
+    return line[0] == ">"
+
+
+def extract_fasta_seq_names(fasta_name: str) -> set:
+    """Extracts the sequence names from a FASTA file.
+
+    modified from Brent Pedersen
+    Correct Way To Parse A Fasta File In Python
+    given a fasta file. yield tuples of header, sequence
+    from https://www.biostars.org/p/710/
+
+    Args:
+      fasta_name: The path to the FASTA file.
+
+    Returns:
+      A set of the sequence names in the FASTA file.
+    """
+
+    # first open the file outside
+    fh = open(fasta_name)
+
+    # ditch the boolean (x[0]) and just keep the header or sequence since
+    # we know they alternate.
+    faiter = (x[1] for x in groupby(fh, is_header))
+
+    for i, header in enumerate(faiter):
+        line = next(header)
+        if is_header(line):
+            # drop the ">"
+            headerStr = line[1:].strip().split()[0]
+        yield headerStr
+
+
+def extract_genes_in_genome(fasta: str, gtf_in: str, gtf_out: str) -> None:
+    """Extracts the genes in the genome from a GTF file.
+
+    Args:
+      fasta: The path to the FASTA file.
+      gtf_in: The path to the input GTF file.
+      gtf_out: The path to the output GTF file.
+
+    Raises:
+      ValueError: If no overlap is found or if the GTF file is not tab delimited.
+    """
+
+    def is_tab_delimited(file):
+        with open(file, "r") as f:
+            return "\t" in f.readline()
+
+    if not is_tab_delimited(gtf_in):
+        raise ValueError("The GTF file is not tab delimited.")
+
+    seq_names_in_genome = set(extract_fasta_seq_names(fasta))
+    logger.info(f"Extracted chromosome sequence names from {fasta}")
+    logger.info("All chromosome names: " + ", ".join(sorted(seq_names_in_genome)))
+
+    with open(gtf_in) as gtf, open(gtf_out, "w") as out:
+        seq_names_in_gtf = {line.split("\t")[0] for line in gtf if line.strip()}
+        overlap = seq_names_in_genome & seq_names_in_gtf
+        if not overlap:
+            raise ValueError("No overlapping scaffolds found.")
+
+        gtf.seek(0)  # Reset file pointer to the start of the file
+        for line in gtf:
+            if line.split("\t")[0] in overlap:
+                out.write(line)
+
+    logger.info(f"Extracted {len(overlap)} matching sequences from {gtf_in} into {gtf_out}")
+    logger.info("All sequence IDs from GTF: " + ", ".join(sorted(seq_names_in_gtf)))
+    logger.info(f"Wrote matching lines to {gtf_out}")
+
+
+def remove_features_without_transcript_id(gtf_in: str, gtf_out: str) -> None:
+    """
+    Removes gene rows with absent or empty transcript_id attributes from a GTF file.
+
+    Args:
+      gtf_in: Path to the input GTF file.
+      gtf_out: The path to the output GTF file.
+    """
+
+    with open(gtf_in, "r") as f_in, open(gtf_out, "w") as f_out:
+        for line in f_in:
+            transcript_id_match = re.search(r'transcript_id "([^"]+)"', line)
+            if transcript_id_match:
+                f_out.write(line)
```

Comment:
Relative to the Fasta files, the GTFs are small. Thus, it is no a big deal to loop over them repeatedly. 

Nonetheless, something like this would do everything with one iteration instead of 3.

```suggestion
 with open(gtf_in) as gtf, open(gtf_out, "w") as out, open(
        gtf_transcript_out, "w"
    ) as out2:
        line_count_all = 0
        line_count_transcript = 0
        for line in gtf:
            if (
                line.split("\t")[0] in seq_names_in_genome
                and line.count("\t") == num_sep
            ):
                out.write(line)
                line_count_all += 1
                if re.search(r'transcript_id "([^"]+)"', line):
                    out2.write(line)
                    line_count_transcript += 1
        if line_count_all == 0:
            raise ValueError("No overlapping scaffolds found.")
```

Reply:
Thanks for this too- done